### ISSUE TITLE
transfer learning with custom dataset has dominant influence

### ISSUE THREAD

**Issue by Guest-01** [NVlabs/stylegan2-ada#109]
First, I'm not sure whether I'm allowed to ask basic usage questions here, so my apologizes if this issue space is only for bug report.

I tried transfer learning with commands below,
```
!python train.py --outdir ./results --snap={snapshot_count} --cfg=11gb-gpu --data=./datasets/{dataset_name} --augpipe={augs} --mirror={mirrored} --mirrory={mirroredY} --metrics={metric_list} --resume={resume_from} --augpipe="bg"
```
resume argument is `ffhq512`

My custom dataset is approx. 250 of blond-female face images.
training took about 8 hrs and logs are below,
```
tick 0     kimg 0.0      time 2m 02s       sec/tick 20.7    sec/kimg 1291.40 maintenance 101.1  gpumem 7.1   augment 0.000
tick 1     kimg 4.0      time 21m 12s      sec/tick 1130.9  sec/kimg 282.71  maintenance 19.2   gpumem 7.1   augment 0.037
tick 2     kimg 8.0      time 40m 04s      sec/tick 1132.3  sec/kimg 283.06  maintenance 0.0    gpumem 7.1   augment 0.074
tick 3     kimg 12.0     time 58m 58s      sec/tick 1133.5  sec/kimg 283.37  maintenance 0.0    gpumem 7.1   augment 0.109
tick 4     kimg 16.0     time 1h 17m 54s   sec/tick 1136.6  sec/kimg 284.14  maintenance 0.0    gpumem 7.1   augment 0.147
tick 5     kimg 20.0     time 1h 37m 00s   sec/tick 1138.1  sec/kimg 284.52  maintenance 7.9    gpumem 7.1   augment 0.182
tick 6     kimg 24.0     time 1h 56m 01s   sec/tick 1140.7  sec/kimg 285.18  maintenance 0.0    gpumem 7.1   augment 0.214
tick 7     kimg 28.0     time 2h 15m 02s   sec/tick 1141.2  sec/kimg 285.30  maintenance 0.0    gpumem 7.1   augment 0.243
tick 8     kimg 32.0     time 2h 34m 04s   sec/tick 1142.1  sec/kimg 285.53  maintenance 0.0    gpumem 7.1   augment 0.273
tick 9     kimg 36.0     time 2h 53m 16s   sec/tick 1144.3  sec/kimg 286.08  maintenance 7.2    gpumem 7.1   augment 0.296
tick 10    kimg 40.0     time 3h 12m 21s   sec/tick 1145.5  sec/kimg 286.38  maintenance 0.0    gpumem 7.1   augment 0.316
tick 11    kimg 44.0     time 3h 31m 26s   sec/tick 1144.6  sec/kimg 286.14  maintenance 0.0    gpumem 7.1   augment 0.338
tick 12    kimg 48.0     time 3h 50m 32s   sec/tick 1146.4  sec/kimg 286.60  maintenance 0.0    gpumem 7.1   augment 0.358
tick 13    kimg 52.0     time 4h 09m 46s   sec/tick 1146.2  sec/kimg 286.55  maintenance 7.0    gpumem 7.1   augment 0.378
tick 14    kimg 56.0     time 4h 28m 53s   sec/tick 1147.2  sec/kimg 286.80  maintenance 0.0    gpumem 7.1   augment 0.393
tick 15    kimg 60.0     time 4h 48m 01s   sec/tick 1148.6  sec/kimg 287.14  maintenance 0.0    gpumem 7.1   augment 0.407
tick 16    kimg 64.0     time 5h 07m 09s   sec/tick 1147.4  sec/kimg 286.86  maintenance 0.0    gpumem 7.1   augment 0.420
tick 17    kimg 68.0     time 5h 26m 24s   sec/tick 1148.2  sec/kimg 287.05  maintenance 7.2    gpumem 7.1   augment 0.435
tick 18    kimg 72.0     time 5h 45m 34s   sec/tick 1149.4  sec/kimg 287.34  maintenance 0.0    gpumem 7.1   augment 0.448
tick 19    kimg 76.0     time 6h 04m 43s   sec/tick 1149.4  sec/kimg 287.34  maintenance 0.0    gpumem 7.1   augment 0.462
tick 20    kimg 80.0     time 6h 23m 54s   sec/tick 1150.7  sec/kimg 287.68  maintenance 0.0    gpumem 7.1   augment 0.473
tick 21    kimg 84.0     time 6h 43m 11s   sec/tick 1150.4  sec/kimg 287.60  maintenance 7.3    gpumem 7.1   augment 0.484
tick 22    kimg 88.0     time 7h 02m 23s   sec/tick 1151.2  sec/kimg 287.80  maintenance 0.0    gpumem 7.1   augment 0.498
tick 23    kimg 92.0     time 7h 21m 34s   sec/tick 1151.3  sec/kimg 287.83  maintenance 0.0    gpumem 7.1   augment 0.510
tick 24    kimg 96.0     time 7h 40m 46s   sec/tick 1152.1  sec/kimg 288.02  maintenance 0.0    gpumem 7.1   augment 0.520
tick 25    kimg 100.0    time 8h 00m 06s   sec/tick 1152.1  sec/kimg 288.02  maintenance 7.4    gpumem 7.1   augment 0.531
tick 26    kimg 104.0    time 8h 19m 18s   sec/tick 1152.8  sec/kimg 288.20  maintenance 0.0    gpumem 7.1   augment 0.540
```

And as result, I compared images from original model(ffhq512) versus custom-trained model
![image](https://user-images.githubusercontent.com/49602144/141257636-b0e05047-f941-4a17-afd6-7999e7a2da5a.png)

My question here is,
although I trained about 8 hrs of my custom datasets, the base model (ffhq512) should have much longer training history.
But how come the result model creates images that are mostly influenced from my custom dataset(blond female)?

Is that just a nature of training? the last given training data deciding final look of images?

**Comment by Kaoru8** [comment]
So, um... what kind of result were you expecting to see instead? It's kind of hard to help you if no one knows what you're trying to achieve in the first place.

The whole concept behind training any kind of machine learning model, whether you're training it from scratch or doing transfer learning, is to "fit it" to a target domain - that is, to modulate its weights in such a way that its outputs most accurately reproduce/classify/whatever-the-task-is items from that domain. In this case, you're training the model to most accurately reproduce photo-realistic portraits.

So, if you train the model on a target domain/dataset that's composed entirely of young blonde females, you're instructing it to synthesize nothing BUT young blonde females - any other output is radically different than the target domain, so it has a low accuracy/high loss value associated with, so the model learns not to do it. Your example shows exactly that happening, ie. it's the expected result.

You really need to state what you're actually expecting to see and trying to achieve so people can help you better, but I can take some guesses and give some general tips...
Let's say you have a model trained on face images. There are some features inherently associated with them, in this case gender (male, female), race (white, black, asian), age (child, adult, old) and hair color (blonde, black) are some obvious examples. These features more than likely weren't evenly distributed in the original dataset the model was trained on, so among other things the model learns to replicate these weights/biases. If the original dataset had 80% white, 10% black and 10% asian samples, you can expect it to output roughly the same proportions of race during synthesis. If 75% of the samples were female, you can expect it to sythesize a female example roughly 75% of the time. This is a vast oversimplification and things aren't quite that linear in practice of course, but the general concept holds.

So, you want to re-train the model to make it produce more outputs that match specific combinations of these features, and less/none of the others. Using our example features from above, we have 2 genders, 3 races, 3 age brackets and 2 hair colors, so 2*3*3*2 = 36 distinct categories any given example can be put in. This is when things get fuzzy, because how you approach things depends on what you want to do, and again, we don't know what that is.

Probably the simplest case, and what the training dataset you used implies you're trying to do, is to generate samples exclusively from one of these 36 categories - in this case, "female - white - adult - blonde". You would then create a dataset composed exclusively of that category and excluding any other combination of features. That's what you did, so that's what you got.

If you instead want to retain most of the original distribution of features, but increase the likelihood of any given example being from the "female - white - adult - blonde" category, you would replace items in the dataset that don't match your target criteria with those that do. So the proportion of your desired examples would grow relative to other categories and the overall model, but the proportions of other categories relative to each other would be kept the same.

If you wanted to generate exclusively "female - white - adult", with a higher than normal chance of blonde hair (let's say 75%), you would create a dataset of exclusively female white adults, where 75% of the samples had blonde hair. Long story short, the composition and distribution of features in your dataset will roughly but directly dictate the composition and distribution of features in the images the model learns to output. If you want non-blondes to be a possible output, you need to have non-blondes in the dataset.

You also seem confused with the concept of transfer learning, and might have thought it means the model will automagically "remember" the initial dataset and just "add" the additional examples to all the original ones (FFHQ) it was trained on, therefore just increasing the distribution of young blondes while leaving the overall appearance of samples unchanged, but that's not how it works at all. When training the model, transfer learning or otherwise, you're telling it "learn to produce these new samples". Transfer learning just means that the model doesn't have to be trained from scratch - instead of learning how to produce a realistic human face from absolute zero (a blank canvas) which is an extremely difficult and computationally expensive task, it can instead just learn the differences between the domain it was already trained on and the target domain. Even if the domains are very different (like re-training a network trained on faces to produce buildings instead) it's still way faster and cheaper than it would be to train from scratch. When the domains are highly similar or nearly identical, like here where you're training a face network on more faces, it's orders of magnitude faster, because the model already knows like 99% of the data required to represent the new faces you're feeding it, so it just needs to learn the tiny differences and unique features present in the new samples.

This is also what lets you get away with using way fewer training samples when re-training vs training from scratch. Try training a GAN from scratch using only 250 samples, and you'll get absolutely nothing as output. Honestly, it's kind of remarkable that the model did such a good job adapting to your target domain with so few samples, even when using transfer learning and in the same domain. But you'll need to drastically increase that number and probably decrease training time regardless of what you're trying to do, because it results in severe overfitting and there's virtually zero diversity in the output, like you can clearly see in your examples - all of them have extremely similar hair styles and faces/facial structure, they're borderline the same person with slight variations rather than a bunch of different people.

Anyways, hope that was of some help at least. If you're still confused or unsure how to proceed just ask and I'll try to help more. But, you know... state what the desired output actually is :D Can't help you reach a goal without first knowing what the goal is.

**Comment by Guest-01** [comment]
My expectations were to get more variations in female blond faces, that is, mixing some features from base model(FFHQ) and my dataset together. But as you noticed, I get very little variations that is mostly within the boundaries of my dataset. So I got many similar faces that has no feature of base model. Your assumption is right, I thought transfer-learning would automatically "remember" the pretrained datasets and "continue" or "add" my dataset. But I get the point. It doesn't work that way. Your explanation was helpful. Thanks a lot!
